DSCS-863: add minLength to all required schema fields, where applicable

sharkySharks opened the pull request:
## Description

While we assert in the product schema that the `sku` field is not null, it should also not be an empty string. Recently ran into some customer issues where an empty sku field left events unprocessed. This schema needs to be updated in the snowplow console, but I will wait for feedback here before making any schema changes. Also interested in the process for updating this schema within snowplow.

## Related Issue

Please see internal adobe jira issue DSCS-863.

## Motivation and Context

`sku` is a required field, which means it should also not be empty.

## How Has This Been Tested?

Downstream processing pipelines for recommendations were breaking because the sku field was empty for some events for some customers. In those pipelines queries that assumed the sku would not be empty on product events have been updated to explicitly remove empty strings as well as the null values so that the pipelines do not error with missing sku information. Twofold, schema validation should also handle empty product event sku information so that events that do not meet our expectations fail fast and return accurate assumptions to our customers about which product information may be missing required sku data.

## Screenshots (if appropriate):

## Types of changes

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

-   [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
-   [x] My code follows the code style of this project.
-   [ ] My change requires a change to the documentation.
-   [x] I have updated the documentation accordingly. - _I believe the schema is self documenting, but happy to update other docs if necessary_
-   [x] I have read the **CONTRIBUTING** document.
-   [ ] I have added tests to cover my changes.
-   [x] All new and existing tests passed.

sharkySharks committed:
DSCS-863: add minLength to product schema

sharkySharks commented on examples/events/snowplow-debugger/product.json:
```diff
@@ -174,4 +175,4 @@
       "version": "2-0-5"
```
I am guessing this is a patch bump to `2-0-6` or would be consider this a minor change to `2-1-0` since this could start rejecting events from stores where the sku field is empty? Also not entirely clear if schema changes are entirely manual and this file is used to reflect what is up in snowplow.

aniham commented:
@sharkySharks I like the idea and glad we are doing this.

The file you're changing here is for documentation only. The correct steps to actually do what you're trying to do are below:

1. Go into the Snowplow console and create a new version of the `Product` schema in QA. `2.1.0` is correct - this is a breaking change that will result in many more events ending up in the Bad Bucket.
2. Test locally against Snowplow QA by changing the schema URL to point to new version:
https://github.com/adobe/commerce-events/blob/main/packages/storefront-events-collector/src/schemas.ts#L9
3. Publish Snowplow changes to Prod
4. Submit PR here that changes docs AND schema file above ^^ to point to new URL

Let me know if any of this is unclear!

sharkySharks committed:
update version for minor version schema change

aniham commented:
@sharkySharks yes for QA you have to be on VPN, sorry forgot to mention earlier

sharkySharks committed:
changed version to patch bump due to snowplow restriction

sharkySharks commented on examples/events/snowplow-debugger/product.json:
```diff
@@ -9,7 +9,8 @@
       },
```
The more I am thinking about this, should we also add `minLength` to the `name` field since it is also required?

krithikachandran replied:
I am not sure why "name" is a required field. I don't think we have any filters related to that in the code today. If we do, we can consider adding minLength otherwise I'd leave it as is

sharkySharks replied:
I think if we are unsure of the effects then I'll just keep the minLength added to `sku` and we can revisit `name` at a different point. I'm not sure we can be sure of all of the ways adding minLength to `name` will change things 😬

sharkySharks commented:
For context on the version number, snowplow did not allow for the minor version to be set to 2-1-0, this is the message that I received and decided to go with the patch bump version:

<img width="776" alt="Screenshot 2024-02-02 at 11 17 15 AM" src="https://github.com/adobe/commerce-events/assets/7784948/1a5c625e-daee-45ec-8236-852601f3d46f">

sharkySharks commented:
Update on testing, I was able to test this out in the qa environment locally, sent productPageView and addToCart events from localhost, saw them end up in the good bucket in kibana. Nothing in the bad bucket from those events. Not sure if other events need to be tested as well. 

I think we should also consider minLength for the required name field as well. 

Also, I do not have production promotion permissions in snowplow as far as I can tell. But latest code here is what I tested against.

aniham commented:
@sharkySharks this sounds right. I think you tested the right things though you should probably do a search through the repo to make sure no other events are using the product context (or if they are, to test those as well).

Next steps will be:

- merge new snowplow schema version to prod (push button)
- merge your code here to QA
- have shirley and vinod run their smoke tests and verify all is good in QA

After that is done, you should be able to publish your change here to prod!

aniham commented:
@sharkySharks note that once this change goes into prod, we're going to see a uptick of events going to the bad bucket - that is the correct behavior and what we want ofc - but worth keeping an eye out and doing some reporting on what changed.

krithikachandran commented:
> For context on the version number, snowplow did not allow for the minor version to be set to 2-1-0, this is the message that I received and decided to go with the patch bump version:
> 
> <img alt="Screenshot 2024-02-02 at 11 17 15 AM" width="776" src="https://private-user-images.githubusercontent.com/7784948/301979278-1a5c625e-daee-45ec-8236-852601f3d46f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcxNTM0NzEsIm5iZiI6MTcwNzE1MzE3MSwicGF0aCI6Ii83Nzg0OTQ4LzMwMTk3OTI3OC0xYTVjNjI1ZS1kYWVlLTQ1ZWMtODIzNi04NTI2MDFmM2Q0NmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDVUMTcxMjUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzM2MzIxM2NkMjY2NjI2ZjAxOTY4MDczNmVmZGQ3MmQzNTY1OWNhMjY5YTY3N2E0NTAyZmJhNzNkNmJmNDY0MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.r8u738SYiAQggOZwmC4-F09rY1Ss8Jj1EnztnxGV4fc">

I think it should be 3-0-0 since it's a breaking change. That way, we can roll it back easily if something goes wrong

sharkySharks commented:
@krithikachandran I am good with using the major version upgrade, particularly if we think it is going to be a breaking change even if it is a bug fix. I will redeploy it out with that version and test. 

@aniham I don't have permissions to push the schema to prod when it is ready. I don't see that button, if you could check my permissions I can do that after I test the major version upgrade.

aniham commented:
@sharkySharks you're now `admin` in Snowplow - should be able to merge!

sharkySharks committed:
bump version to major for breaking change

sharkySharks commented:
Ran both positive and negative tests against the new schema 3-0-0.

Here is what happens in the negative case:

<img width="1247" alt="Screenshot 2024-02-06 at 6 03 27 PM" src="https://github.com/adobe/commerce-events/assets/7784948/ff0f2f3a-2bf2-486f-94bf-b08f1800eafd">

migrated the schema to production in snowplow. 

<img width="1422" alt="Screenshot 2024-02-06 at 6 06 01 PM" src="https://github.com/adobe/commerce-events/assets/7784948/99aea2b9-5ebb-4d45-9232-87267a8150d8">

If all looks good to you here I'll wait for a final sign off on the latest schema version update on this pr and then go through QA  testing tomorrow for the app.

sharkySharks committed:
update required fields with minLength

sharkySharks committed:
update schema.ts

sharkySharks commented:
Hey folks, I've updated this pr, I have not tested this yet as I want to get feedback on the potential changes first and then move forward with making the changes in snowplow for qa and run through all the testing paths once there is agreement on these changes.

I went ahead and added `minLength` to required fields for `string` and `array` fields that did not also include `null` in their accepted types. I have not done the work to verify what fields should still be required or not, I just blanketed all required string and array fields with `minLength` for schemas in this repo. 

This is coming out of numerous outages this year where required fields were empty and broke various parts of ML pipeline processing.

priyasbhat commented on examples/events/snowplow-debugger/search-input.json:
```diff
@@ -11,15 +12,17 @@
       "query": {
-        "type": "string"
+        "type": "string",
+        "minLength": 1
```
I think this maps to `phrase` in `productSearch` request. If so, there are cases where it could be empty (when browsing a category page, gql query filters on a category but no phrase is provided).

krithikachandran commented on examples/events/snowplow-debugger/recommended-item.json:
```diff
@@ -54,7 +58,8 @@
                   "properties": {
                     "code": {
-                      "type": "string"
+                      "type": "string",
```
I have rarely seen finalAdjustments being set in price data. Doesn't look like it's marked as required, we can remove this

krithikachandran commented on examples/events/snowplow-debugger/search-result-product.json:
```diff
@@ -2,19 +2,23 @@
       "name": {
-        "type": "string"
+        "type": "string",
+        "minLength": 1
```
What happens here if the search returns no results? We might still need the events but with empty fields?

grahamcrackers commented:
@sharkySharks is this still needed?

sharkySharks commented:
let's close for now, it was a good idea at the time but may be a correction that causes more problems. if it comes up again we can reassess.
